File: VB/Mapper/Core/MapperExpressionVisitor.vb
Imports System.Linq.Expressions
Imports System.Collections.Generic

Namespace Core
    Friend Class MapperExpressionVisitor
        Inherits ExpressionVisitor

#Region "Variables"

        Private previousMember As Expression

        Private ReadOnly checkNull As Boolean

        Private ReadOnly paramSource As ParameterExpression

        Private ReadOnly membersTocheck As List(Of MemberExpression)

#End Region

#Region "Constructeur"

        ''' <summary>
        ''' Initialise une nouvelle instance de  <see cref="MapperExpressionVisitor"/> classe.
        ''' </summary>
        ''' <param name="checkIfNull">Indique si l'on tester la nullité des objets (récursive)</param>
        ''' <param name="paramClassSource">paramètre de la source</param>
        Friend Sub New(checkIfNull As Boolean, paramClassSource As ParameterExpression)
            checkNull = checkIfNull
            paramSource = paramClassSource
            membersTocheck = New List(Of MemberExpression)()
        End Sub

#End Region

#Region "Méthodes surchargées"

        ''' <summary>
        ''' Distribue l'expression à l'une des méthodes de visite les plus spécialisées dans cette classe.
        ''' </summary>
        ''' <param name="node">Expression à visiter.</param>
        ''' <returns>
        ''' Expression modifiée, si celle-ci ou toute sous-expression a été modifiée ; sinon, retourne l'expression d'origine.
        ''' </returns>
        Public Overrides Function Visit(node As Expression) As Expression

            If checkNull Then
                'On traite que nos cas
                Select Case node.NodeType
                    Case ExpressionType.MemberAccess
                        Return Me.VisitMember(TryCast(node, MemberExpression))
                    Case ExpressionType.Parameter
                        Return Me.VisitParameter(TryCast(node, ParameterExpression))
                    Case ExpressionType.Convert
                        Return Me.VisitMember(TryCast(TryCast(node, UnaryExpression).Operand, MemberExpression))
                    Case ExpressionType.Lambda
                        MyBase.Visit((TryCast(node, LambdaExpression)).Body)
                        Exit Select
                    Case Else
                        MyBase.Visit(node)
                        Exit Select
                End Select
                Dim isFirst As Boolean = True
                'Nous voulons tester tout les sous objets avant d'affecter la valeur
                'Ex:
                'Expression d'origine : 
                'source.SubClass.SubClass2.MaPropriete
                'Expression transformée :
                'If(source.SubClass <> Nothing, If(source.SubClass.SubClass2 <> Nothing, source.SubClass.SubClass2.MaPropriete, Nothing), Nothing)
                For Each item As MemberExpression In membersTocheck

                    If Not isFirst Then 'pour ne pas tester la valeur à affecter

                        Dim defaultValue As Object = GetDefaultValue(item.Type)

                        'Création de la vérification de la nullité
                        Dim notNull As Expression = Expression.NotEqual(item, Expression.Constant(defaultValue, item.Type))
                        Dim conditional As Expression = Nothing
                        'On crée une condition qui inclue la condition précédente
                        If previousMember IsNot Nothing Then
                            Dim defaultPreviousValue As Object = GetDefaultValue(previousMember.Type)
                            conditional = Expression.Condition(notNull, previousMember, Expression.Constant(defaultPreviousValue, previousMember.Type))
                        End If
                        'On affecte la condition nouvellement crée qui deviendra la précédente
                        previousMember = conditional
                    Else 'là, la propriété demandée
                        previousMember = item
                        isFirst = False
                    End If
                Next
                Return previousMember
            Else
                'retour par défaut (avec changement du paramètre)
                'pour supprimer la validation de l'expression lambda
                If (node.NodeType = ExpressionType.Lambda) Then
                    Return MyBase.Visit((TryCast(node, LambdaExpression)).Body)
                Else
                    Return MyBase.Visit(node)
                End If
            End If
        End Function

        ''' <summary>
        ''' Visite <see cref="T:System.Linq.Expressions.ParameterExpression" />.
        ''' </summary>
        ''' <param name="node">Expression à visiter.</param>
        ''' <returns>
        ''' Expression modifiée, si celle-ci ou toute sous-expression a été modifiée ; sinon, retourne l'expression d'origine.
        ''' </returns>
        Protected Overrides Function VisitParameter(node As ParameterExpression) As Expression
            'Pour changer de paramètre
            Return paramSource
        End Function

        ''' <summary>
        ''' Visite les enfants de <see cref="T:System.Linq.Expressions.MemberExpression" />.
        ''' </summary>
        ''' <param name="node">Expression à visiter.</param>
        ''' <returns>
        ''' Expression modifiée, si celle-ci ou toute sous-expression a été modifiée ; sinon, retourne l'expression d'origine.
        ''' </returns>
        Protected Overrides Function VisitMember(node As MemberExpression) As Expression
            Dim memberAccessExpression As MemberExpression = DirectCast(MyBase.VisitMember(node), MemberExpression)
            'Pour traiter plus tard
            If memberAccessExpression IsNot Nothing AndAlso checkNull Then
                'Sachant que l'on visite le premier membre en premier et que l'on descend à chaque fois
                'on insert notre membre courant à la première ligne de la liste pour inverser l'ordre
                'exemple :
                'source.SubClass.SubClass2.MaPropriete
                'la liste serait avec Add:
                'MaList(0) = SubClass
                'MaList(1) = SubClass2
                'MaList(2) = MaPropriete

                'mais nous voulons
                'MaList(0) = MaPropriete
                'MaList(1) = SubClass2
                'MaList(2) = SubClass
                membersTocheck.Insert(0, memberAccessExpression)
            End If
            Return memberAccessExpression
        End Function

        ''' <summary>
        ''' Visite les enfants de <see cref="T:System.Linq.Expressions.UnaryExpression" />.
        ''' </summary>
        ''' <param name="node">Expression à visiter.</param>
        ''' <returns>
        ''' Expression modifiée, si celle-ci ou toute sous-expression a été modifiée ; sinon, retourne l'expression d'origine.
        ''' </returns>
        Protected Overrides Function VisitUnary(node As UnaryExpression) As Expression
            Return Me.VisitMember(TryCast(node.Operand, MemberExpression))
        End Function

        Private Function GetDefaultValue(typeObject As Type) As Object
            Dim defaultValue As Object = Nothing
            'Dans le cas de type valeur (ex :Integer), il faut instancier l'objet pour avoir sa valeur par défaut
            If typeObject.BaseType = GetType(ValueType) Then
                Dim exp = Expression.[New](typeObject)
                Dim lambda = Expression.Lambda(exp)
                Dim constructor = lambda.Compile()
                defaultValue = constructor.DynamicInvoke()
            End If
            Return defaultValue
        End Function


#End Region

    End Class
End Namespace

File: CS/MapperCore/Core/MapperExpressionVisitor.cs
using System.Collections.Generic;
using System.Linq.Expressions;
using System;
using MapperExpression.Helper;

namespace MapperExpression.Core
{
    internal class MapperExpressionVisitor : ExpressionVisitor
    {
        #region Variables


        private bool checkNull;

        private readonly ParameterExpression parameter;

        private readonly Stack<MemberExpression> membersToCheck;

        internal ParameterExpression Parameter
        {
            get
            {
                return parameter;
            }
        }
        #endregion

        #region Constructor


        internal MapperExpressionVisitor(ParameterExpression paramClassSource)
        {

            parameter = paramClassSource;
            membersToCheck = new Stack<MemberExpression>();
        }

        #endregion

        #region Overloaded methods


        /// <summary>
        /// Distributes the expression one of the more specialized visit methods in this class.
        /// </summary>
        /// <param name="node">Expression to visit.</param>
        /// <param name="checkIfNullity">Check</param>
        /// <returns>
        /// Altered expression, if it or any subexpression was modified; otherwise, returns the original expression.
        /// </returns>
        public Expression Visit(Expression node, bool checkIfNullity = false)
        {
            checkNull = checkIfNullity;
            Expression result = null;
            if (node == null)
                return node;
            if (checkNull)
            {
                // Treated our case
                switch (node.NodeType)
                {
                    case ExpressionType.MemberAccess:
                        result = VisitMember(node as MemberExpression);
                        break;
                    case ExpressionType.Parameter:
                        result = VisitParameter(node as ParameterExpression);
                        break;
                    case ExpressionType.Convert:
                        result = VisitMember((node as UnaryExpression).Operand as MemberExpression);
                        break;
                    case ExpressionType.Lambda:
                        // to remove validation of the lambda expression
                        result = base.Visit((node as LambdaExpression).Body);
                        break;
                    default:
                        result = base.Visit(node);
                        break;

                }
                bool isFirst = true;
                Expression previousExpression = null;
                if (membersToCheck.Count > 1)
                {
                    // We want to test all the sub objects before assigning the value
                    // Ex: source.SubClass.SubClass2.MyProperty
                    // Which will give
                    // source.SubClass != null ? source.SubClass.SubClass2 != null ? source.SubClass.SubClass2.MyProperty :DefaultValueOfProperty :DefaultValueOfProperty
                    foreach (MemberExpression item in membersToCheck)
                    {

                        if (!isFirst) // Not to test the value of the property back
                        {
                            object defaultValue = MapperHelper.GetDefaultValue(item.Type);

                            // Creating verification of default value
                            Expression notDefaultValue = Expression.NotEqual(item, Expression.Constant(defaultValue, item.Type));
                            Expression conditional = null;
                            // It creates a condition that includes the above condition
                            if (previousExpression != null)
                            {
                                object defaultPreviousValue = MapperHelper.GetDefaultValue(previousExpression.Type);
                                conditional = Expression.Condition(notDefaultValue, previousExpression, Expression.Constant(defaultPreviousValue, previousExpression.Type));
                            }
                            // It affects the newly created conditions that will become the previous
                            previousExpression = conditional;
                        }
                        else // here the requested property
                        {
                            previousExpression = item;
                            isFirst = false;
                        }
                    }

                    return previousExpression;
                }
                // For one element don't need recusrive
                else if (membersToCheck.Count == 1)
                {
                    var item = membersToCheck.Peek();
                    object defaultValue = MapperHelper.GetDefaultValue(item.Type);
                    // Creating verification of default value
                    Expression notDefaultValue = Expression.NotEqual(item, Expression.Constant(defaultValue, item.Type));
                    Expression conditional = null;
                    conditional = Expression.Condition(notDefaultValue, item, Expression.Constant(defaultValue, item.Type));

                    return conditional;
                }
                return result;

            }
            else
            {
                // return by default (with change of the parameter)
                // to remove validation of the lambda expression
                if ((node.NodeType == ExpressionType.Lambda))
                {
                    return base.Visit(((LambdaExpression)node).Body);
                }
                else
                {
                    return base.Visit(node);
                }
            }
        }

        /// <summary>
        /// Visit <see cref="T:System.Linq.Expressions.ParameterExpression" />.
        /// </summary>
        /// <param name="node">Expression to visit.</param>
        /// <returns>
        /// Altered expression, if it or any subexpression was modified; otherwise, returns the original expression.
        /// </returns>
        protected override Expression VisitParameter(ParameterExpression node)
        {
            // To change parameter
            return parameter;
        }

        /// <summary>
        /// Visit the children of <see cref="T:System.Linq.Expressions.MemberExpression" />.
        /// </summary>
        /// <param name="node">Expression to visit.</param>
        /// <returns>
        /// Altered expression, if it or any subexpression was modified; otherwise, returns the original expression.
        /// </returns>
        protected override Expression VisitMember(MemberExpression node)
        {
            if (node == null)
            {
                return node;
            }
            MemberExpression memberAccessExpression = (MemberExpression)base.VisitMember(node);

            // To treat later
            if (memberAccessExpression != null && checkNull)
            {
                // Knowing that the last member is in the first visit and we go back every time
                // our current insert member is in the first line of the list to change the order
                membersToCheck.Push(memberAccessExpression);
            }
            return memberAccessExpression;
        }

        /// <summary>
        ///  Visit the children of <see cref="T:System.Linq.Expressions.UnaryExpression" />.
        /// </summary>
        /// <param name="node">Expression to visit.</param>
        /// <returns>
        /// Altered expression, if it or any subexpression was modified; otherwise, returns the original expression.
        /// </returns>
        protected override Expression VisitUnary(UnaryExpression node)
        {
            if (node != null)
            {
                if (node.Operand.NodeType == ExpressionType.MemberAccess)
                {
                    return VisitMember(node.Operand as MemberExpression);
                }
                if (node.NodeType == ExpressionType.Convert)
                {
                    return Visit(node.Operand);
                }
            }
            return node;
        }

        #endregion


    }
}